ndinterp/src/grid.rs

```diff
@@ -49,19 +55,28 @@ impl Grid {
         let dy_b = self.derivative_at(index);
         0.5 * (dy_f + dy_b)
     }
+}
 
+impl<D: Dimension> Grid<D> {
     /// Find the index of the last value in the input such that input(idx) < query
     /// If the query is outside the grid returns an extrapolation error
-    pub fn closest_below(&self, query: f64) -> Result<usize, InterpolationError> {
-        if query > self.input[self.input.len() - 1] {
-            Err(InterpolationError::ExtrapolationAbove(query))
-        } else if query < self.input[0] {
-            Err(InterpolationError::ExtrapolationBelow(query))
-        } else {
-            let u_idx = self.input.iter().position(|x| x > &query).unwrap();
-            let idx = u_idx - 1;
-            Ok(idx)
+    pub fn closest_below<const N: usize>(
+        &self,
+        input_query: &[f64],
+    ) -> Result<[usize; N], InterpolationError> {
```

Comment:
Here, I believe you'd like to use something like `D::NDIM`.

Unfortunately, `D::NDIM: Option<usize>`, so you'd need something like `.unwrap()`, but executed at compile time.

https://docs.rs/ndarray/0.15.6/ndarray/trait.Dimension.html#associatedconstant.NDIM

Reply:
A simpler alternative might be to split the bulk of the function, and then implement thin wrappers around (one per `Dimension`).
Even the implementors of `Dimension` itself are just 7 (+ 1).

Reply:
Otherwise, you should be able to use a const function to unwrap it, defining an associated constant, and try to use the associated constant in the signature.

https://users.rust-lang.org/t/compile-time-const-unwrapping/51619

EDIT: Unfortunately, const panic is not stabilized, because of an issue with Rust 2021 https://github.com/rust-lang/rust/issues/85194, maybe you could check if const pattern matching is allowed

Reply:
A const function to unwrap IxN at compile time you mean?

Reply:
Indeed, in order to turn it into a `usize`.

Reply:
I think this cannot be done. Or at least I cannot find a way of passing the `D` to the const function:

```
generic parameters may not be used in const operations
type parameter may not be used in const expressions
```

and

```
const parameters may only be used as standalone arguments
```

Not sure whether that means there's no way of doing what I want to do (without creating  a path per dimension, at that point I rather leave the `N`)

Reply:
I'd really like to solve this issue. But if it's not simple, we could postpone.